src/GridFS/Exception/StreamException.php

```diff
@@ -0,0 +1,40 @@
+<?php
+
+namespace MongoDB\GridFS\Exception;
+
+use MongoDB\Exception\RuntimeException;
+use function MongoDB\BSON\fromPHP;
+use function MongoDB\BSON\toJSON;
+use function sprintf;
+use function stream_get_meta_data;
+
+class StreamException extends RuntimeException
+{
+    /** @param resource $destination */
+    public static function downloadFromFilenameFailed(string $filename, $destination) : self
+    {
+        $metadata = stream_get_meta_data($destination);
+
+        return new static(sprintf('Downloading GridFS file "%s" to resource "%s" failed: stream error.', $filename, $metadata['uri']));
+    }
+
+    /**
+     * @param mixed    $id
+     * @param resource $destination
+     */
+    public static function downloadFromIdFailed($id, $destination) : self
+    {
+        $stringId = toJSON(fromPHP(['_id' => $id]));
```

Comment:
This logic may lead to more concise (and consistent) output: https://github.com/mongodb/mongo-php-library/blob/b3a4c2412d38f709babdf7d54f51914062eb49bc/src/GridFS/Bucket.php#L632-L636

Although, if you end up using the error messages I suggested above, this might not even be necessary if you decide "GridFS identifier" is redundant since it'd already be reported in the `<gridfs-uri>`.

Reply:
Now that you mention it, that logic contains a flaw: passing an array for `_id` will result in the string "Array" being used, since it triggers the first check (it's not an object). I'll investigate the effect of using this logic in `createPathForFile` separately, as using an identifier like `['foo' => 'bar']` is technically fine.